From cc78f1251c43d0d5592d08aa8edac584cad7a329 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:41 -0400
Subject: [powerpc] numa: Enable CONFIG_HAVE_MEMORYLESS_NODES

Message-id: <1410545655-205645-533-git-send-email-dzickus@redhat.com>
Patchwork-id: 94643
O-Subject: [RHEL7 PATCH 532/626] powerpc/numa: Enable CONFIG_HAVE_MEMORYLESS_NODES
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 64bb80d87f01ec01c76863b61b457e0904387f2f
Author: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
Date:   Fri May 16 16:41:20 2014 -0700

    powerpc/numa: Enable CONFIG_HAVE_MEMORYLESS_NODES

    Based off fd1197f1 for ia64, enable CONFIG_HAVE_MEMORYLESS_NODES if
    NUMA. Initialize the local memory node in start_secondary.

    With this commit and the preceding to enable
    CONFIG_USER_PERCPU_NUMA_NODE_ID, which is a prerequisite, in a PowerKVM
    guest with the following topology:

    numactl --hardware
    available: 3 nodes (0-2)
    node 0 cpus: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22
    23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46
    47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70
    71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94
    95 96 97 98 99
    node 0 size: 1998 MB
    node 0 free: 521 MB
    node 1 cpus: 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114
    115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132
    133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150
    151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168
    169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186
    187 188 189 190 191 192 193 194 195 196 197 198 199
    node 1 size: 0 MB
    node 1 free: 0 MB
    node 2 cpus:
    node 2 size: 2039 MB
    node 2 free: 1739 MB
    node distances:
    node   0   1   2
      0:  10  40  40
      1:  40  10  40
      2:  40  40  10

    the unreclaimable slab is reduced by close to 130M:

    Before:
            Slab:             418176 kB
            SReclaimable:      26624 kB
            SUnreclaim:       391552 kB

    After:
            Slab:             298944 kB
            SReclaimable:      31744 kB
            SUnreclaim:       267200 kB

    Signed-off-by: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/Kconfig b/arch/powerpc/Kconfig
index 42511c2..97a7e48 100644
--- a/arch/powerpc/Kconfig
+++ b/arch/powerpc/Kconfig
@@ -459,6 +459,10 @@ config USE_PERCPU_NUMA_NODE_ID
  def_bool y
  depends on NUMA
 
+config HAVE_MEMORYLESS_NODES
+ def_bool y
+ depends on NUMA
+
 config ARCH_SELECT_MEMORY_MODEL
  def_bool y
  depends on PPC64
diff --git a/arch/powerpc/kernel/smp.c b/arch/powerpc/kernel/smp.c
index d7252ad..4863ea1 100644
--- a/arch/powerpc/kernel/smp.c
+++ b/arch/powerpc/kernel/smp.c
@@ -755,6 +755,7 @@ void start_secondary(void *unused)
   * numa_node_id() works after this.
   */
  set_numa_node(numa_cpu_lookup_table[cpu]);
+ set_numa_mem(local_memory_node(numa_cpu_lookup_table[cpu]));
 
  smp_wmb();
  notify_cpu_starting(cpu);
-- 
1.7.1